src/libraries/System.Net.Sockets/tests/FunctionalTests/TelemetryTest.cs

```diff
@@ -65,36 +365,117 @@ public void EventSource_EventsRaisedAsExpected()
-
-                    VerifyEventCounter("incoming-connections-established", eventCounters);
-                    VerifyEventCounter("outgoing-connections-established", eventCounters);
-                    VerifyEventCounter("bytes-received", eventCounters);
-                    VerifyEventCounter("bytes-sent", eventCounters);
-                    VerifyEventCounter("datagrams-received", eventCounters);
-                    VerifyEventCounter("datagrams-sent", eventCounters);
+
+                    VerifyStartStopEvents(events, connect: true, expectedCount: 10);
+
+                    Assert.DoesNotContain(events, e => e.EventName == "ConnectFailed");
+                    Assert.DoesNotContain(events, e => e.EventName == "ConnectCanceled");
+
+                    VerifyEventCounters(events, connectCount: 10, shouldHaveTransferedBytes: true, shouldHaveDatagrams: true);
                 }
             }).Dispose();
         }
 
-        private static void VerifyEvents(IEnumerable<EventWrittenEventArgs> events, string eventName, int expectedCount)
+        private static void VerifyStartStopEvents(ConcurrentQueue<EventWrittenEventArgs> events, bool connect, int expectedCount)
         {
-            EventWrittenEventArgs[] starts = events.Where(e => e.EventName == eventName).ToArray();
+            string startName = connect ? "ConnectStart" : "AcceptStart";
+            EventWrittenEventArgs[] starts = events.Where(e => e.EventName == startName).ToArray();
             Assert.Equal(expectedCount, starts.Length);
+            foreach (EventWrittenEventArgs start in starts)
+            {
+                object startPayload = Assert.Single(start.Payload);
+                Assert.False(string.IsNullOrWhiteSpace(startPayload as string));
+            }
+
+            string stopName = connect ? "ConnectStop" : "AcceptStop";
+            EventWrittenEventArgs[] stops = events.Where(e => e.EventName == stopName).ToArray();
+            Assert.Equal(expectedCount, stops.Length);
+            Assert.All(stops, stop => Assert.Empty(stop.Payload));
+
+            for (int i = 0; i < expectedCount; i++)
+            {
+                Assert.NotEqual(Guid.Empty, starts[i].ActivityId);
+                Assert.Equal(starts[i].ActivityId, stops[i].ActivityId);
+            }
         }
 
-        private static void VerifyEventCounter(string name, Dictionary<string, double> eventCounters)
+        private static async Task WaitForEventAsync(ConcurrentQueue<EventWrittenEventArgs> events, string name)
         {
-            Assert.True(eventCounters.ContainsKey(name));
-            Assert.True(eventCounters[name] > 0);
+            DateTime startTime = DateTime.UtcNow;
+            while (!events.Any(e => e.EventName == name))
+            {
+                if (DateTime.UtcNow.Subtract(startTime) > TimeSpan.FromSeconds(30))
+                    throw new TimeoutException($"Timed out waiting for {name}");
+
+                await Task.Delay(100);
```

Comment:
Wouldn't a shorter delay work better here? Don't we react too late in some of the test cases, if eg. start events are not queued immediately?

Reply:
You mean decreasing the `Delay(100)` here?

Reply:
Yes.

Reply:
I don't think it's wrong as we're just waiting for the event to appear in the collection, but we could reduce the time taken by the test by a few ms.